lib/backend/firestore/firestorebk.go

```diff
@@ -0,0 +1,716 @@
+	_, err = docSnap.Ref.Update(ctx, updates)
+	if err != nil {
+		return ConvertGRPCError(err)
+	}
+	return err
+}
+
+// Close closes the Firestore client contexts and releases associated resources
+func (b *FirestoreBackend) Close() error {
+	b.clientCancel()
+	err := b.buf.Close()
+	if err != nil {
+		b.Logger.Error("error closing buffer, continuing with closure of other resources...", err)
+	}
+	return b.svc.Close()
+}
+
+// CloseWatchers closes all the watchers without closing the backend
+func (b *FirestoreBackend) CloseWatchers() {
+	b.buf.Reset()
+}
+
+// Clock returns wall clock
+func (b *FirestoreBackend) Clock() clockwork.Clock {
+	return b.clock
+}
+
+// newLease
+func (b *FirestoreBackend) newLease(item backend.Item) *backend.Lease {
+	var lease backend.Lease
+	if item.Expires.IsZero() {
+		return &lease
+	}
+	lease.Key = item.Key
+	return &lease
+}
+
+// hashKey hashes keys using SHA1 enabling fast lookups
+func (b *FirestoreBackend) hashKey(key []byte) string {
+	h := sha1.New()
+	h.Write(key)
+	sha1Bytes := h.Sum(nil)
+	return fmt.Sprintf("%x", sha1Bytes)
+}
+
+// RetryingAsyncFunctionRunner wraps a task target in retry logic
+func RetryingAsyncFunctionRunner(ctx context.Context, retryConfig utils.LinearConfig, logger *log.Logger, task func() error, taskName string) {
+	retry, err := utils.NewLinear(retryConfig)
+	if err != nil {
+		logger.Errorf("bad retry parameters: %v, returning and not running", err)
+		return
+	}
+	for {
+		err := task()
+		if err != nil && (err != context.Canceled || status.Convert(err).Code() != codes.Canceled) {
+			logger.Errorf("%v returned with error: %v", taskName, err)
+		}
+		logger.Debugf("reloading %v for %s", retry, taskName)
+		select {
+		case <-retry.After():
+			retry.Inc()
+		case <-ctx.Done():
+			logger.Debugf("returning from %v loop", taskName)
+			return
+		}
+	}
+}
+
+// watchCollection watches a firestore collection for changes and pushes those changes, events into the buffer for watchers
+func (b *FirestoreBackend) watchCollection() error {
+	snaps := b.svc.Collection(b.CollectionName).Snapshots(b.clientContext)
+	b.signalWatchStart()
+	defer snaps.Stop()
+	for {
+		querySnap, err := snaps.Next()
+		if err == context.Canceled {
+			return nil
+		} else if err != nil {
+			return ConvertGRPCError(err)
+		}
+		for _, change := range querySnap.Changes {
+			var r record
+			err = change.Doc.DataTo(&r)
+			if err != nil {
+				return ConvertGRPCError(err)
+			}
+			var expires time.Time
+			if r.Expires != 0 {
+				expires = time.Unix(r.Expires, 0)
+			}
+			var e backend.Event
+			switch change.Kind {
+			case firestore.DocumentAdded:
+				fallthrough
+			case firestore.DocumentModified:
+				e = backend.Event{
+					Type: backend.OpPut,
+					Item: backend.Item{
+						Key: []byte(r.Key),
+						Value: []byte(r.Value),
+						Expires: expires,
+						ID: r.ID,
```

Comment:
Research the revision property, can we use some incremental unique doc id from firestore?